wallet-new/src/Cardano/Wallet/Kernel/DB/Sqlite.hs

```diff
@@ -0,0 +1,562 @@
+{-# LANGUAGE TupleSections      #-}
+{-# OPTIONS_GHC -fno-warn-orphans #-}
+
+-- | Sqlite database for the 'TxMeta' portion of the wallet kernel.
+module Cardano.Wallet.Kernel.DB.Sqlite (
+
+    -- * Resource acquisition
+      openMetaDB
+    , closeMetaDB
+    , MetaDBHandle
+
+    -- * Basic API
+    , putTxMeta
+    , getTxMeta
+    , getTxMetas
+
+    -- * Filtering and sorting primitives
+    , Limit (..)
+    , Offset (..)
+    , Sorting (..)
+    , SortCriteria (..)
+    , SortDirection (..)
+
+    -- * Unsafe functions
+    , unsafeMigrate
+    ) where
+
+import qualified Prelude
+import           Universum
+
+import           Database.Beam.Backend.SQL (FromBackendRow, HasSqlValueSyntax (..),
+                                            IsSql92DataTypeSyntax, intType, varCharType)
+import           Database.Beam.Backend.SQL.SQL92 (Sql92OrderingExpressionSyntax,
+                                                  Sql92SelectOrderingSyntax)
+import           Database.Beam.Query (HasSqlEqualityCheck, (==.))
+import qualified Database.Beam.Query as SQL
+import qualified Database.Beam.Query.Internal as SQL
+import           Database.Beam.Schema (Beamable, Database, DatabaseSettings, PrimaryKey, Table)
+import qualified Database.Beam.Schema as Beam
+import           Database.Beam.Sqlite.Connection (Sqlite, runBeamSqlite)
+import           Database.Beam.Sqlite.Syntax (SqliteCommandSyntax, SqliteDataTypeSyntax,
+                                              SqliteExpressionSyntax, SqliteSelectSyntax,
+                                              SqliteValueSyntax, fromSqliteCommand,
+                                              sqliteRenderSyntaxScript)
+import qualified Database.SQLite.Simple as Sqlite
+import           Database.SQLite.Simple.FromField (FromField (..), returnError)
+import qualified Database.SQLite.SimpleErrors as Sqlite
+import qualified Database.SQLite.SimpleErrors.Types as Sqlite
+
+import           Control.Exception (throwIO, toException)
+import           Control.Lens (Getter)
+import qualified Data.Foldable as Foldable
+import qualified Data.Map as M
+import           Data.Time.Units (fromMicroseconds, toMicroseconds)
+import           Database.Beam.Migrate (CheckedDatabaseSettings, DataType (..), Migration,
+                                        MigrationSteps, boolean, createTable, evaluateDatabase,
+                                        executeMigration, field, migrationStep, notNull,
+                                        runMigrationSteps, unCheckDatabase, unique)
+import           Formatting (build, formatToString, sformat, (%))
+import           GHC.Generics (Generic)
+
+import qualified Cardano.Wallet.Kernel.DB.TxMeta.Types as Kernel
+import qualified Pos.Core as Core
+import           Pos.Crypto.Hashing (decodeAbstractHash, hashHexF)
+
+
+-- | An opaque handle to the underlying 'Sqlite.Connection'.
+data MetaDBHandle = MetaDBHandle {
+    _mDbHandleConnection :: Sqlite.Connection
+}
```

Comment:
in the interest of easier testing/mocking, I think I would rather see this as a "record of functions" that contain the API we're implementing :smile: Then we can swap SQLite out for a pure database if we ever need to for testing.